pkg/certificate/auto/provider.go

```diff
@@ -0,0 +1,295 @@
+	if err != nil {
+		if k8serrors.IsNotFound(err) {
+			return err
+		} else {
+			return fmt.Errorf("invalid certificate secret: %s present in namespace: %s, please delete and try again.\nError: %s",
+				secretName, namespace, err)
+		}
+	}
+
+	log.Printf("Valid certificate secret: %s already present in namespace: %s", secretName, namespace)
+	return nil
+}
+
+func (ac *AutoCertProvider) ValidateCertificateSecret(ctx context.Context, secretName, namespace string,
+	_ *interfaces.Config) error {
+	var err error
+	secret := &corev1.Secret{}
+	err = ac.Get(ctx, client.ObjectKey{Name: secretName, Namespace: namespace}, secret)
+	if err != nil && k8serrors.IsNotFound(err) {
+		for try := range interfaces.MaxRetries {
+			// Wait for the certificate secret to get created
+			log.Printf("Valid certificate secret: retry attempt %v, after %v, error: %v", try+1, interfaces.RetryInterval, err)
+			time.Sleep(interfaces.RetryInterval)
+			err = ac.Get(ctx, client.ObjectKey{Name: secretName, Namespace: namespace}, secret)
+			if err == nil {
+				break
+			}
+		}
+		if err != nil {
+			return err
+		}
+	} else {
+		return err
+	}
+
+	certificateData, exists := secret.Data["tls.crt"]
+	if !exists {
+		return interfaces.ErrTLSCert
+	}
+
+	decodeCertificatePem, _ := pem.Decode(certificateData)
+	if decodeCertificatePem == nil {
+		return interfaces.ErrDecodeCert
+	}
+
+	privateKeyData, keyExists := secret.Data["tls.key"]
+	if !keyExists {
+		return interfaces.ErrTLSKey
+	}
+
+	parseCert, err := x509.ParseCertificate(decodeCertificatePem.Bytes)
+	if err != nil {
+		return fmt.Errorf("failed to parse certificate: %w", err)
+	}
+
+	if parseCert.NotAfter.Before(time.Now()) {
+		return interfaces.ErrCertExpired
+	}
+
+	privateKey, err := parsePrivateKey(privateKeyData)
+	if err != nil {
+		return fmt.Errorf("failed to parse private key: %w", err)
+	}
+
+	if checkKeyPairErr := checkKeyPair(parseCert, privateKey); checkKeyPairErr != nil {
+		return fmt.Errorf("private key does not match certificate: %w", checkKeyPairErr)
+	}
+
+	return nil
+}
+
+func (ac *AutoCertProvider) DeleteCertificateSecret(ctx context.Context, secretName, namespace string) error {
+	var secret corev1.Secret
+	err := ac.Get(ctx, client.ObjectKey{Name: secretName, Namespace: namespace}, &secret)
+	if k8serrors.IsNotFound(err) {
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+	return ac.Delete(ctx, &secret)
+}
+
+func (ac *AutoCertProvider) RevokeCertificate(ctx context.Context, secretName string, namespace string) error {
+	return ac.DeleteCertificateSecret(ctx, secretName, namespace)
+}
+
+func (ac *AutoCertProvider) GetCertificateConfig(ctx context.Context,
+	secretName, namespace string) (*interfaces.Config, error) {
+	var autoCertSecret corev1.Secret
+	err := ac.Get(ctx, client.ObjectKey{Name: secretName, Namespace: namespace}, &autoCertSecret)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get certificate: %w", err)
+	}
+
+	return &interfaces.Config{}, nil
```

Comment:
We should get the info from the certificate included in the secret.

Reply:
Hi @ahrtr , I explored getting the original config from the secret itself but it will be a complicated one since we are manipulating the config and passing it to `transport.SelfCert()` which in turn creats the certificate.
Another workaround is to retrieve the certificateConfig from the etcdCluster object via the `certificateSecret.ownerRef.Name`, but it will cause a circular dependency

What can be the suggested solution here?
cc @ivanvc